File: c/complex_text/alternate_font_lists/alternate_fonts.c
/* ============================================================================
 *  alternate_fonts -- C (x64) port of the VB6 mirror
 *  examples\Vb6\complex_text\alternate_font_lists\alternate_fonts.bas
 *  Complex text layout with an alternate font list to improve font
 *  substitution. Multi-language text loaded as UTF-16 and laid out with the
 *  wide version of WriteFTextEx().
 * ========================================================================== */
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>
#include "lumaspdf.h"

#define TXT_FILE "E:\\LUMASPDFSDK\\examples\\test_files\\multi_lang.txt"

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType)
{
    (void)Data; (void)ErrCode; (void)ErrType;
    printf("%s\n", ErrMessage ? ErrMessage : "");
    return 0;   /* We try to continue if an error occurs */
}

static void exedir(const char* a0, char* out, size_t n)
{
    char* s;
    strncpy(out, a0, n - 1); out[n - 1] = 0;
    s = strrchr(out, '\\'); if (!s) s = strrchr(out, '/');
    if (s) *s = 0; else strcpy(out, ".");
}

/* Reads a file raw as UTF-16 (2 bytes per char), NUL terminated. */
static LWCHAR* GetFileBuffer(const char* FileName)
{
    FILE* f; long n; char* b;
    f = fopen(FileName, "rb");
    if (!f) return NULL;
    fseek(f, 0, SEEK_END); n = ftell(f); fseek(f, 0, SEEK_SET);
    if (n < 0) { fclose(f); return NULL; }
    b = (char*)malloc(n + 2);
    if (!b) { fclose(f); return NULL; }
    if (n > 0) fread(b, 1, n, f);
    fclose(f);
    b[n] = 0; b[n + 1] = 0;
    return (LWCHAR*)b;
}

int main(int argc, char** argv)
{
    PPDF pdf;
    LWCHAR* txt;
    SI32 altFonts;
    char dir[1024], outFile[1200];
    /* Alternate fonts, sorted alphabetically. */
    LWCHAR* fonts[5] = {
        L"Malgun Gothic",   /* Korean */
        L"Mangal",          /* Hindi or Marathi */
        L"Nyala",           /* Amharic */
        L"Shonar Bangla",   /* Bengali */
        L"Shruti"           /* Gujarati */
    };

    exedir(argv[0], dir, sizeof(dir));

    pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, ErrProc);
    pdfCreateNewPDFA(pdf, "");           /* We create no PDF file in this example */

    txt = GetFileBuffer(TXT_FILE);

    pdfSetPageCoords(pdf, pcTopDown);
    /* Enable complex text layout */
    pdfSetGStateFlags(pdf, gfComplexText, 0);

    altFonts = pdfCreateAltFontList(pdf);
    pdfSetAltFontsW(pdf, altFonts, fonts, 5);

    pdfAppend(pdf);

    /* The font must be loaded with cpUnicode. */
    pdfSetFontA(pdf, "Arial", fsRegular, 10.0, 1, cpUnicode);
    /* Activate the alternate font list */
    pdfActivateAltFontList(pdf, altFonts, 1);

    pdfSetLeading(pdf, pdfGetTypoLeading(pdf));
    pdfWriteFTextExW(pdf, 50.0, 50.0,
                     pdfGetPageWidth(pdf) - 100.0,
                     pdfGetPageHeight(pdf) - 100.0, taJustify, txt);

    free(txt);
    pdfEndPage(pdf);

    outFile[0] = 0;
    if (pdfHaveOpenDoc(pdf) != 0) {
        sprintf(outFile, "%s\\out.pdf", dir);
        if (pdfOpenOutputFileA(pdf, outFile) == 0) {
            pdfDeletePDF(pdf);
            return 0;
        }
    }
    if (pdfCloseFile(pdf) != 0)
        printf("PDF file \"%s\" successfully created!\n", outFile);

    pdfDeletePDF(pdf);
    return 0;
}

File: csharp/complex_text/alternate_font_lists/alternate_fonts.cs
//  alternate_fonts -- C# port of
//  examples\Vb6\complex_text\alternate_font_lists\alternate_fonts.bas
//  Complex text layout with an alternate font list to improve font substitution.
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using LumasPdfSdk;

class AlternateFonts
{
    static TErrorProc _err = ErrProc;
    static int ErrProc(IntPtr Data, int ErrCode, string ErrMessage, int ErrType)
    {
        Console.WriteLine(ErrMessage);
        return 0;
    }

    static string GetFileBuffer(string fileName)
    {
        try { return Encoding.Unicode.GetString(File.ReadAllBytes(fileName)); }
        catch { return ""; }
    }

    static void Main()
    {
        string[] fonts = {
            "Malgun Gothic",   // Korean
            "Mangal",          // Hindi or Marathi
            "Nyala",           // Amharic
            "Shonar Bangla",   // Bengali
            "Shruti"           // Gujarati
        };

        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        LumasPdf.pdfCreateNewPDFW(pdf, "");

        string txt = GetFileBuffer(Path.GetFullPath(Path.Combine(
            AppDomain.CurrentDomain.BaseDirectory, "../../../test_files/multi_lang.txt")));

        LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown);
        LumasPdf.pdfSetGStateFlags(pdf, LumasPdfConsts.gfComplexText, false);

        // Provide an alternate font list (array of PWideChar).
        int altFonts = LumasPdf.pdfCreateAltFontList(pdf);
        IntPtr[] ptrs = new IntPtr[fonts.Length];
        for (int i = 0; i < fonts.Length; i++)
            ptrs[i] = Marshal.StringToHGlobalUni(fonts[i]);

        GCHandle h = GCHandle.Alloc(ptrs, GCHandleType.Pinned);
        LumasPdf.pdfSetAltFontsW(pdf, (uint)altFonts, h.AddrOfPinnedObject(), (uint)fonts.Length);
        h.Free();

        LumasPdf.pdfAppend(pdf);
        LumasPdf.pdfSetFontW(pdf, "Arial", LumasPdfConsts.fsRegular, 10.0, true, TCodepage.cpUnicode);
        LumasPdf.pdfActivateAltFontList(pdf, altFonts, true);

        LumasPdf.pdfSetLeading(pdf, LumasPdf.pdfGetTypoLeading(pdf));
        LumasPdf.pdfWriteFTextExW(pdf, 50.0, 50.0,
            LumasPdf.pdfGetPageWidth(pdf) - 100.0, LumasPdf.pdfGetPageHeight(pdf) - 100.0,
            (int)LumasPdfConsts.taJustify, txt);
        LumasPdf.pdfEndPage(pdf);

        foreach (IntPtr p in ptrs) Marshal.FreeHGlobal(p);

        string outFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf");
        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile))
            {
                LumasPdf.pdfDeletePDF(pdf);
                return;
            }
        }
        if (LumasPdf.pdfCloseFile(pdf))
            Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");
        LumasPdf.pdfDeletePDF(pdf);
    }
}
